lib/pure/times.nim

```diff
@@ -489,6 +425,102 @@ template lqImpl(a: Duration|Time, b: Duration|Time): bool =
 template eqImpl(a: Duration|Time, b: Duration|Time): bool =
   a.seconds == b.seconds and a.nanosecond == b.nanosecond
 
+proc initDuration*(nanoseconds, microseconds, milliseconds,
+                   seconds, minutes, hours, days, weeks: int64 = 0): Duration =
+  runnableExamples:
+    let dur = initDuration(seconds = 1, milliseconds = 1)
+    doAssert dur.milliseconds == 1
+    doAssert dur.seconds == 1
+
+  let seconds = convert(Weeks, Seconds, weeks) +
+    convert(Days, Seconds, days) +
+    convert(Minutes, Seconds, minutes) +
+    convert(Hours, Seconds, hours) +
+    convert(Seconds, Seconds, seconds) +
+    convert(Milliseconds, Seconds, milliseconds) +
+    convert(Microseconds, Seconds, microseconds) +
+    convert(Nanoseconds, Seconds, nanoseconds)
+  let nanoseconds = (convert(Milliseconds, Nanoseconds, milliseconds mod 1000) +
+    convert(Microseconds, Nanoseconds, microseconds mod 1_000_000) +
+    nanoseconds mod 1_000_000_000).int
+  # Nanoseconds might be negative so we must normalize.
+  result = normalize[Duration](seconds, nanoseconds)
+
+const DurationZero* = initDuration() ## \
+  ## Zero value for durations. Useful for comparisons.
+  ##
+  ## .. code-block:: nim
+  ##
+  ##   doAssert initDuration(seconds = 1) > DurationZero
+  ##   doAssert initDuration(seconds = 0) == DurationZero
+
+proc toDurationParts*(dur: Duration): DurationParts =
+  ## Converts a duration into an array consisting of fixed time units.
+  ##
+  ## Each value in the array gives information about a specific unit of
+  ## time, for example ``result[Days]`` gives a count of days.
+  ##
+  ## This procedure is useful for converting ``Duration`` values to strings.
+  runnableExamples:
+    let dp = toDurationParts(initDuration(weeks=2, days=1))
+    doAssert dp[Days] == 1
+    doAssert dp[Weeks] == 2
+  var remS = dur.seconds
+  var remNs = dur.nanosecond.int
+
+  # Normally ``nanoseconds`` should always be positive, but
+  # if it's not, ensure it
+  if remNs < 0:
+    remNs -= convert(Seconds, Nanoseconds, 1)
+    remS.inc 1
+
+  for unit in countdown(Weeks, Seconds):
+    let quantity = convert(Seconds, unit, remS)
+    remS = remS mod convert(unit, Seconds, 1)
+
+    result[unit] = quantity
+
+  for unit in countdown(Milliseconds, Nanoseconds):
+    let quantity = convert(Nanoseconds, unit, remNs)
+    remNs = remNs mod convert(unit, Nanoseconds, 1)
+
+    result[unit] = quantity
+
+
+proc `$`*(dur: Duration): string =
+  ## Human friendly string representation of ``dur``.
+  runnableExamples:
+    doAssert $initDuration(seconds = 2) == "2 seconds"
+    doAssert $initDuration(weeks = 1, days = 2) == "1 week and 2 days"
+    doAssert $initDuration(hours = 1, minutes = 2, seconds = 3) == "1 hour, 2 minutes, and 3 seconds"
+    doAssert $initDuration(milliseconds = -1500) == "-1 second and -500 milliseconds"
```

Comment:
Nanoseconds should not be forced to be positivt, the sign of all duration parts should be the same. I think it's the comment in the code that is wrong.

Reply:
```nim
  # Ensure the same sign for seconds and nanoseconds
  if remS < 0 and remNs > 0:
    remNs -= convert(Seconds, Nanoseconds, 1)
    remS.inc 1
  elif remS > 0 and remNs < 0:
    remNs += convert(Seconds, Nanoseconds, 1)
    remS.dec 1
```

Reply:
I think that's the right way to do it

Reply:
`dur.nanoseconds` can never be negative since it has the type `NanosecondRange`,  so you don't need to check for it. I think the original code was correct.

Reply:
Thanks for the comment. Fixed

Reply:
Sorry for the confusion i created about the negative nanoseconds. But now 
```nim
echo initDuration(days = -1)
```
gives (again):
`-23 hours, -59 minutes, -59 seconds, and -1000 milliseconds`

I think, the better solution would be
```nim
  # Ensure the same sign for seconds and nanoseconds
  if remS < 0 and remNs > 0:
    remNs -= convert(Seconds, Nanoseconds, 1)
    remS.inc 1
```

Reply:
done

Reply:
better solution, in my opinion, is a strict check for 0